### PR TITLE
model for census division level prices

### PR DESCRIPTION
**Pull request by andrew-sha**
a new model in stahl to associate each store with a census division and province, and modified the product_timeseries_metrics model in wolfram to get aggregate statistics at the census division level, also new migration files for the postgis extension

### CODE REVIEW

File: ganymede/models/sources.yml [hunk 1]
```diff
@@ -10,4 +10,12 @@ sources:
       - name: retailers
       - name: stores
       - name: statcan_food_prices
+  - name: static_resources
```
Comment (thecartercodes) [inline ganymede/models/sources.yml]:
```suggestion
  - name: static
```

File: ganymede/models/sources.yml [hunk 2]
```diff
@@ -10,4 +10,12 @@ sources:
...
+    schema: public
+    tables:
+      - name: censusdivisions
```
Comment (thecartercodes) [inline ganymede/models/sources.yml]:
```suggestion
      - name: census_divisions
```

Let's lower snake case everything for consistency.

File: ganymede/models/stahl/cd_locate_stores.sql [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{{ config(materialized='view') }}
```
Comment (thecartercodes) [inline ganymede/models/stahl/cd_locate_stores.sql]: let's name models as objects as opposed to phrases, functions... you can call this `dim_stores` if you want. It's popular naming convention. read more here - https://stackoverflow.com/questions/1716220/star-schema-naming-conventions

File: ganymede/models/stahl/cd_locate_stores.sql [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+
+with store_locations as (
+    select id, r_store_attributes->'geo'->'latitude' as latitude, r_store_attributes->'geo'->'longitude' as longitude, {{ raw_region_to_province("store_address->>'addressRegion'") }} as region
+    from {{ source('aethervest', 'stores') }}
+)
+select store_locations.id, cds.cdname, store_locations.region 
+from store_locations inner join {{ source('static_resources', 'censusdivisions') }} cds
```
Comment (thecartercodes) [inline ganymede/models/stahl/cd_locate_stores.sql]:
```suggestion
from store_locations inner join {{ source('static', 'census_divisions') }} cds
```

If you change this to a left join, you can then make an assertion that cdname is not null in the test

File: ganymede/models/wolfram/product_timeseries_metrics.sql [hunk 5]
```diff
@@ -21,15 +22,16 @@ with all_calendar_dates as (
   from all_calendar_dates acd
   left join {{ source('aethervest', 'product_listings_history') }} plh
     on acd.val > plh.effective_from and acd.val <= coalesce(plh.effective_to, '9999-01-01'::timestamp)
-  join {{ source('aethervest', 'stores')}} s
+  join {{ ref('cd_locate_stores')}} s
```
Comment (thecartercodes) [inline ganymede/models/wolfram/product_timeseries_metrics.sql]: this is good here.

File: ganymede/models/stahl/cd_locate_stores.sql (hunk 3 above)
Comment (thecartercodes) [reply ganymede/models/stahl/cd_locate_stores.sql]: Let's use this as a dim table for stores where we denormalize all dimensions of the store and anticipate its use case in downstream wolfram resources. I.e. let's flatten and expand all the columns and alias the names.

**Review by thecartercodes — CHANGES_REQUESTED** [review]
Looks good

**Comment by thecartercodes** [comment]
We also discussed this in the dms, go ahead and add a makefile command for the static import of the census divisions.

File: ganymede/models/stahl/cd_locate_stores.sql (hunk 4 above)
Comment (andrew-sha) [reply ganymede/models/stahl/cd_locate_stores.sql]: Just so I understand (kinda new to the fancier kinds of sql joins)--using a left join we can ensure that if for some reason a store cannot be located within a census division then that store record will have null in its cdname column?

Comment (thecartercodes) [reply ganymede/models/stahl/cd_locate_stores.sql]: Yes, and this way, we can write something like this in the dbt test - 

```
models:
  - name: dim_store
    columns:
      - name: census_division_name
        tests:
          - not_null
```

Speaking of which, you should add the model to the `.yml` ... I generally tend to not use acronyms if possible btw e.g. (cd_name)  ... especially if the resource is to be bound to the api, which this prolly will. This way, there's minimum relabeling in the API and the end-user blind to this part of the repo has some understanding of the definitions without having to read to many docs.

Comment (andrew-sha) [reply ganymede/models/stahl/cd_locate_stores.sql]: Got it

Seems like there currently is no `stahl.yml` file, should I make one?

File: ganymede/models/stahl/cd_locate_stores.sql [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+on ST_Covers(cds.geom, ST_Transform(ST_SetSRID(ST_MakePoint(longitude::double precision, latitude::double precision), 4326), 3347))
```
Comment (thecartercodes) [inline ganymede/models/stahl/cd_locate_stores.sql]: Can you use words to describe to me what's happening with this function here? New to me as well.

Comment (thecartercodes) [reply ganymede/models/stahl/cd_locate_stores.sql]: I don't really follow the these `srid` integer values... scanned the docs and I'm guessing they pertain to some coordinate system ::shrugs::?

**Review by thecartercodes — COMMENTED** [review]
I read this and now have another question cuz I don't know what these `srid` constants are.

File: Makefile [hunk 7]
```diff
@@ -16,3 +16,6 @@ bootstrap-runner-prod-remote: scripts/bootstrap.sh
 		 export SCOPULI_RUNNER_PROD_HOST=${SCOPULI_RUNNER_PROD_HOST}; \
 		 export SCOPULI_RUNNER_ENV=prod; \
 		 bash -s" < scripts/bootstrap.sh
+
+load-static:
+	shell scripts/load_static.sh
```
Comment (thecartercodes) [inline Makefile]:
```suggestion
	bash scripts/load_static.sh
```

File: ganymede/models/sources.yml [hunk 8]
```diff
@@ -10,4 +10,12 @@ sources:
...
       - name: stores
       - name: statcan_food_prices
+  - name: static
+    schema: public
+    tables:
+      - name: census_divisions
+  - name: stahl
+    schema: stahl
+    tables:
+      - name: dim_stores
```
Comment (thecartercodes) [inline ganymede/models/sources.yml]:
```suggestion
      - name: dim_store
```

File: ganymede/models/stahl/dim_stores.sql [hunk 9]
```diff
@@ -0,0 +1,21 @@
+{{ config(materialized='view') }}
```
Comment (thecartercodes) [inline ganymede/models/stahl/dim_stores.sql]: Rename this to `dim_store`

File: ganymede/models/stahl/dim_stores.sql [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+with store_coordinates as (
+    select *, r_store_attributes->'geo'->'latitude' as latitude, r_store_attributes->'geo'->'longitude' as longitude, {{ raw_region_to_province("store_address->>'addressRegion'") }} as region
```
Comment (thecartercodes) [inline ganymede/models/stahl/dim_stores.sql]:
```suggestion
    select *, r_store_attributes->'geo'->'latitude' as latitude, r_store_attributes->'geo'->'longitude' as longitude, {{ raw_region_to_province("store_address->>'addressRegion'") }} as region_code
```

File: ganymede/models/stahl/dim_stores.sql [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+select 
+    store_coordinates.id
+    , store_coordinates.retailer_name
+    , store_coordinates.parent_retailer_name
+    , store_coordinates.store_address
+    , store_coordinates.version
+    , store_coordinates.r_id
+    , store_coordinates.r_store_attributes
+    , cds.cdname as census_division_name 
+    , store_coordinates.region
```
Comment (thecartercodes) [inline ganymede/models/stahl/dim_stores.sql]:
```suggestion
    , store_coordinates.region_code
```

File: ganymede/models/wolfram/product_timeseries_metrics.sql [hunk 12]
```diff
@@ -21,15 +22,16 @@ with all_calendar_dates as (
   from all_calendar_dates acd
   left join {{ source('aethervest', 'product_listings_history') }} plh
     on acd.val > plh.effective_from and acd.val <= coalesce(plh.effective_to, '9999-01-01'::timestamp)
-  join {{ source('aethervest', 'stores')}} s
+  join {{ ref('dim_stores')}} s
```
Comment (thecartercodes) [inline ganymede/models/wolfram/product_timeseries_metrics.sql]:
```suggestion
  join {{ ref('dim_store')}} s
```

File: scripts/load_static.sh [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+psql -h $SCOPULI_PROD_PG_HOST -p $SCOPULI__PROD_PG_PORT -d $SCOPULI_PROD_PG_DBNAME -U $SCOPULI_PROD_PG_USERNAME < migrations/20231010_importpostgis.up.sql
+
+cd_filename=tmp_lcd_000b21a_e.zip
+
+curl -o $cd_filename https://www12.statcan.gc.ca/census-recensement/2021/geo/sip-pis/boundary-limites/files-fichiers/lcd_000b21a_e.zip
+
+unzip $cd_filename -d tmp/
+rm -f $cd_filename
+
+shp2pgsql -D -I -s 3347 tmp/lcd_000b21a_e.shp static.statcan_census_divisions | psql -h $SCOPULI_PROD_PG_HOST -p $SCOPULI__PROD_PG_PORT -d $SCOPULI_PROD_PG_DBNAME -U $SCOPULI_PROD_PG_USERNAME
```
Comment (thecartercodes) [inline scripts/load_static.sh]:
```suggestion
shp2pgsql -D -I -s 3347 tmp/lcd_000b21a_e.shp static.statcan_census_divisions | psql -h $SCOPULI_PROD_PG_HOST -p $SCOPULI_PROD_PG_PORT -d $SCOPULI_PROD_PG_DBNAME -U $SCOPULI_PROD_PG_USERNAME
```

Test that this works with local env vars.

File: scripts/load_static.sh [hunk 14]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+export PGPASSWORD=$SCOPULI_PROD_PG_PASSWORD
```
Comment (thecartercodes) [inline scripts/load_static.sh]:
```suggestion
export PGPASSWORD=$SCOPULI_PG_PASSWORD
```

After this to that a more generic variable name... you can then call

```
make load-census-divisions PG_PASSWORD=$SCOPULI_LOCAL_PG_PASSWORD
```

**Review by thecartercodes — CHANGES_REQUESTED** [review]

**Comment by thecartercodes** [comment]
Try to lint this tho

File: ganymede/models/sources.yml [hunk 15]
```diff
@@ -10,4 +10,12 @@ sources:
       - name: retailers
       - name: stores
       - name: statcan_food_prices
+  - name: static
```
Comment (thecartercodes) [inline ganymede/models/sources.yml:14]: You need to resolve a merge conflict here. I think you accidentally deleted a line.

**Review by thecartercodes — APPROVED** [review]
lgtm

**Pull request merged**